Jjo/add sample functions jjo repo

jjo opened the pull request:
mirroring prometheus/prometheus#12503

jjo committed:
Implement sample_limit(N, metric) and sample_ratio(R, metric):

* `sample_limit(N, metric)` returns the 1st N samples gathered from metric, _without_ any particular order
* `sample_ratio(R, metric)` with `0 <= R <= 1.0` returns approximately `R` ratio of samples

Sometimes you want to "sample" some timeseries, for example to explore
their labels names and values, currently the only operator available is
`topk(K, metric)` (and `bottomk() of course), needing to "pay" for its
ordering feature by visiting all samples, and sorting the topk-s.

This change introduces `sample_limit(N, metric)` which _just_ samples
the first points visited without any further/guaranteed ordering, and
also `sample_ratio(R, metric)` for cases where you want a more "fair"
distribution of the samples, using `R` as a "proxy" for an approximate
sampling ratio, which uses `Metrics.Hash()/MaxUint64` as the sample
ratio, and select the sample if its ratio is lesser than `R`.

An `promql/engine.go` excerpt to expand on `sample_ratio()`:
(NOTE: `sampleOffset()` returns `sample.Metric.Hash()) / MaxUint64`)
```
    switch {
    case ratioLimit >= 0:
        // If ratioLimit >= 0: add sample if ratiosampler.sampleOffset() is lesser than ratioLimit
        //
        // 0.0        ratioLimit                1.0
        //  [---------|--------------------------]
        //  [#########...........................]
        //
        // e.g.:
        //   ratiosampler.sampleOffset()==0.3 && ratioLimit==0.4
        //     0.3 < 0.4 ? --> add sample
        //
        if sampleOffset < ratioLimit {
            heap.Push(h, sample)
        }
    case ratioLimit < 0:
        // If ratioLimit < 0: add sample if rand() return the "complement" of ratioLimit>=0 case
        // (loosely similar behavior to negative array index in other programming languages)
        //
        // 0.0       1+ratioLimit               1.0
        //  [---------|--------------------------]
        //  [.........###########################]
        //
        // e.g.:
        //   ratiosampler.sampleOffset()==0.3 && ratioLimit==-0.6
        //     0.3 >= 0.4 ? --> don't add sample
        if sampleOffset >= (1.0 + ratioLimit) {
            heap.Push(h, sample)
        }
    }
```

Testing
-------
* added unit tests to promql/engine_test.go
* added tests to promql/bench_test.go

Web
---
* added web/ui/ artifacts

NB: `sample_ratio()` was previously `sample_random()`, changed to a deterministic implementation, as per PR discussion.

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
add HashRandomizer as a non-random :) deterministic [0.0, 0.1] generator, return value as `labels.Hash() / MaxUint64`

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
fix comment

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
s/sample_random/sample_ratio/, adjust docu and tests

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
no-change small code reorg

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
TestSample: few sample_ratio() adjustments

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
TestSample: increase test case #timeseries

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
Rebase 2023-07-28

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

michalbiesek committed:
Update Go version

Update build/test to use Go 1.21.

Signed-off-by: Michal Biesek <michalbiesek@gmail.com>

michalbiesek committed:
lint: Fix `SA1019` Using a deprecated function

`rand.Read` has been deprecated since Go 1.20
`crypto/rand.Read` is more appropriate

Ref: https://tip.golang.org/doc/go1.20

Signed-off-by: Michal Biesek <michalbiesek@gmail.com>

jjo committed:
Rebase 2023-08-22, adapt to engine_test.go to newer testing interface(s)

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo committed:
Rebase 2023-08-25

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Analysis

- 🎯 **Main theme:** Addition of new sample_limit and sample_ratio functions to Prometheus
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** True
- ✨ **Focused PR:** True
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

-  **General suggestions:** The PR is well-structured and follows good coding practices. The new functions are well-tested and the code is clean and easy to understand. The PR also includes updates to the documentation, which is a good practice when introducing new features.

- 🤖 **Code feedback:**


  - **relevant file:** ``promql/engine.go``
   **suggestion:** Consider adding more comments to explain the logic behind the sample_ratio function, especially the case when ratioLimit < 0. This would make the code easier to understand for other developers. [medium]
   **relevant line:** [case parser.SAMPLE_RATIO:](https://github.com/jjo/prometheus/pull/1/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR2818)


  - **relevant file:** ``promql/engine_test.go``
   **suggestion:** It would be beneficial to add more edge test cases for the new functions, such as what happens when the input parameters are null, or when the parameters are out of expected range. [medium]
   **relevant line:** [func TestSample(t *testing.T) {](https://github.com/jjo/prometheus/pull/1/files#diff-d9232498b406c019b9418d1cf73d477c61f68b55679366152e96970bc1890972R1637)


  - **relevant file:** ``promql/engine.go``
   **suggestion:** It would be beneficial to handle the case where the ratioLimit is exactly -1.0 in the sample_ratio function. Currently, the code only handles the case where ratioLimit is less than -1.0 or greater than 1.0. [medium]
   **relevant line:** [case q < -1.0:](https://github.com/jjo/prometheus/pull/1/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR2553)


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

jjo commented:
@CodiumAI-Agent /improve

QodoAI-Agent commented on promql/engine_test.go:
```diff
@@ -1634,6 +1634,217 @@ load 1ms
+	for v := 0.0; v < 1.0; v += 0.1 {
+		// OR-ing both must return the full matrix
+		ratioCases = append(ratioCases, ratioCase{
+			query: fmt.Sprintf(orQuery, v, -(1.0 - v)),
+			start: 0, end: 20, interval: 10,
+			result: fullMatrix20,
+		})
+		// AND-ing both must return an empty matrix
+		ratioCases = append(ratioCases, ratioCase{
+			query: fmt.Sprintf(andQuery, v, -(1.0 - v)),
+			start: 0, end: 20, interval: 10,
+			result: Matrix{},
+		})
+	}
+
+	for _, c := range ratioCases {
+		t.Run(c.query, func(t *testing.T) {
+			if c.interval == 0 {
+				c.interval = 1
+			}
+			start, end, interval := time.Unix(c.start, 0), time.Unix(c.end, 0), time.Duration(c.interval)*time.Second
+			var err error
+			var qry Query
+			if c.end == 0 {
+				qry, err = engine.NewInstantQuery(context.Background(), storage, nil, c.query, start)
+			} else {
+				qry, err = engine.NewRangeQuery(context.Background(), storage, nil, c.query, start, end, interval)
+			}
+			require.NoError(t, err)
+
+			res := qry.Exec(context.Background())
+			require.NoError(t, res.Err)
+			if expMat, ok := c.result.(Matrix); ok {
+				sort.Sort(expMat)
+				sort.Sort(res.Value.(Matrix))
+			}
+			require.Equal(t, c.result, res.Value, "query %q failed", c.query)
+		})
+	}
+}
```
**Suggestion:** Consider refactoring the test cases for the `sample_limit` and `sample_ratio` functions into separate test functions. This will improve the readability and maintainability of the tests.
```suggestion
func TestSampleLimit(t *testing.T) {
	engine := newTestEngine()
	storage := LoadedStorage(t, `
load 10s
  metric{job="1"} 0+1x1000
  metric{job="2"} 0+2x1000
  metric{job="3"} 0+3x1000
  metric{job="4"} 0+4x1000
  metric{job="5"} 0+5x1000
`)
	t.Cleanup(func() { storage.Close() })
	...
	cases := []struct {
		query                string
		start, end, interval int64 // Time in seconds.
		result               parser.Value
		resultIn             parser.Value
		resultLen            int // Required if resultIn is set
	}{
		...
	}
	...
	for _, c := range cases {
		t.Run(c.query, func(t *testing.T) {
			...
		})
	}
}

func TestSampleRatio(t *testing.T) {
	engine := newTestEngine()
	storage := LoadedStorage(t, `
load 10s
  metric{job="1"} 0+1x1000
  metric{job="2"} 0+2x1000
  metric{job="3"} 0+3x1000
  metric{job="4"} 0+4x1000
  metric{job="5"} 0+5x1000
`)
	t.Cleanup(func() { storage.Close() })
	...
	cases := []struct {
		query                string
		start, end, interval int64 // Time in seconds.
		result               parser.Value
		resultIn             parser.Value
		resultLen            int // Required if resultIn is set
	}{
		...
	}
	...
	for _, c := range cases {
		t.Run(c.query, func(t *testing.T) {
			...
		})
	}
}
```

jjo committed:
split TestSample() cases into TestSample{Limit,Ratio,RatioDynamic}(), improve some comments

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Analysis

- 🎯 **Main theme:** Adding new sample_limit and sample_ratio operators to Prometheus
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** True
- ✨ **Focused PR:** True
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

-  **General suggestions:** The PR is well-structured and includes relevant tests for the new features. It is recommended to ensure that the new operators are documented in the user-facing documentation, and that any potential edge cases are handled correctly.

- 🤖 **Code feedback:**


  - **relevant file:** ``promql/engine_test.go``
   **suggestion:** Consider adding more edge cases for the sample_limit and sample_ratio functions. For example, what happens if N is larger than the number of samples in the metric for sample_limit, or if P is not within the range [0, 1] for sample_ratio? [important]
   **relevant line:** [func TestSampleLimit(t *testing.T) {](https://github.com/jjo/prometheus/pull/1/files#diff-d9232498b406c019b9418d1cf73d477c61f68b55679366152e96970bc1890972R1660)


  - **relevant file:** ``promql/parser/lex.go``
   **suggestion:** It would be good to add comments explaining the new operators, similar to the existing ones. This will help other developers understand the purpose and usage of these operators. [medium]
   **relevant line:** ["sample_limit": SAMPLE_LIMIT,](https://github.com/jjo/prometheus/pull/1/files#diff-0bd1c19fe329f5a2900abfa8fcf023c68afa4e05dae16958a372fd9e350c839aR120)


  - **relevant file:** ``promql/parser/parse.go``
   **suggestion:** It would be beneficial to add error handling for invalid parameters for the new operators. For example, if the limit for sample_limit is negative or if the ratio for sample_ratio is not within the range [0, 1]. [important]
   **relevant line:** [if n.Op == TOPK || n.Op == BOTTOMK || n.Op == QUANTILE || n.Op == SAMPLE_LIMIT || n.Op == SAMPLE_RATIO {](https://github.com/jjo/prometheus/pull/1/files#diff-0accfa26d7e8bfecc27f0c77f5f7ecc5d2417cf16edfa7b8ca9d9ab5b822aff0R502)


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

jjo committed:
add errored test cases

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
@CodiumAI-Agent /review

QodoAI-Agent commented:
## PR Analysis

- 🎯 **Main theme:** Adding new sample_limit() and sample_ratio() operators to Prometheus
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** True
- ✨ **Focused PR:** True
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

-  **General suggestions:** The PR is well structured and includes relevant tests for the new functionality. It also updates the documentation to reflect the changes. However, it would be beneficial to add more comments in the code to explain the logic behind the new operators.

- 🤖 **Code feedback:**


  - **relevant file:** ``promql/engine_test.go``
   **suggestion:** Consider adding more edge cases to your tests to ensure the robustness of your new operators. For example, you could test what happens when the limit or ratio parameters are null or undefined. [important]
   **relevant line:** [func TestSampleLimit(t *testing.T) {](https://github.com/jjo/prometheus/pull/1/files#diff-d9232498b406c019b9418d1cf73d477c61f68b55679366152e96970bc1890972R1660)


  - **relevant file:** ``promql/engine_test.go``
   **suggestion:** It would be beneficial to add more comments in your test cases to explain what each case is testing specifically. This will make it easier for other developers to understand the purpose of each test. [medium]
   **relevant line:** [func TestSampleRatio(t *testing.T) {](https://github.com/jjo/prometheus/pull/1/files#diff-d9232498b406c019b9418d1cf73d477c61f68b55679366152e96970bc1890972R1757)


  - **relevant file:** ``promql/parser/lex.go``
   **suggestion:** It would be beneficial to add comments to explain the purpose of the new operators in the context of the lexer. This will make it easier for other developers to understand their role. [medium]
   **relevant line:** ["sample_limit": SAMPLE_LIMIT,](https://github.com/jjo/prometheus/pull/1/files#diff-0bd1c19fe329f5a2900abfa8fcf023c68afa4e05dae16958a372fd9e350c839aR120)


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

jjo commented:
@CodiumAI-Agent /review -i

QodoAI-Agent commented:
## Incremental PR Review

-  **⏮️ Review for commits since previous PR-Agent review:** Starting from commit https://github.com/jjo/prometheus/pull/1/commits/None
## PR Analysis

- 🎯 **Main theme:** Adding new sampling functions to the Prometheus project
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** True
- ✨ **Focused PR:** Yes, all changes are related to the addition of new sampling functions.
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

-  **General suggestions:** The PR seems well-structured and focused. The commit messages are clear and informative. It would be beneficial to include more context or a use case in the PR description to help reviewers understand the motivation behind these changes.

### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.

jjo commented:
@CodiumAI-Agent /describe

zeitlinger committed:
Unit tests for native histograms (#12668)

promql: Extend testing framework to support native histograms

This includes both the internal testing framework as well as the rules unit test feature of promtool.

This also adds a bunch of basic tests. Many of the code level tests can now be converted to tests within the framework, and more tests can be added easily.

---------

Signed-off-by: Harold Dost <h.dost@criteo.com>
Signed-off-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>
Signed-off-by: Stephen Lang <stephen.lang@grafana.com>
Co-authored-by: Harold Dost <h.dost@criteo.com>
Co-authored-by: Stephen Lang <stephen.lang@grafana.com>
Co-authored-by: Gregor Zeitlinger <gregor.zeitlinger@grafana.com>

SuperQ committed:
Update nodejs version

Update to the latest nodejs LTS release. The old v16 ends security
support 2023.09.11.

* https://nodejs.org/en/blog/announcements/v20-release-announce
* https://nodejs.org/en/blog/announcements/nodejs16-eol

Signed-off-by: SuperQ <superq@gmail.com>

bboreham committed:
promql: force mmap of head chunks in BenchmarkRangeQuery

Otherwise we have a highly unusual situation of over 100 chunks
in the headChunks list of each series, which heavily skews
performance.

Signed-off-by: Bryan Boreham <bjboreham@gmail.com>

bboreham committed:
Merge pull request #12690 from michalbiesek/feat-go-bump

Update Go version to 1.21

bboreham committed:
Merge pull request #12755 from bboreham/rangequery-benchmark-mmap

promql: force mmap of head chunks in BenchmarkRangeQuery

SuperQ committed:
Merge pull request #12754 from prometheus/superq/bump_node

Update nodejs version

jjo committed:
Rebase 2023-08-28

Signed-off-by: JuanJo Ciarlante <juanjosec@gmail.com>

jjo commented:
@CodiumAI-Agent /review -i

QodoAI-Agent commented:
## Incremental PR Review

-  **⏮️ Review for commits since previous PR-Agent review:** Starting from commit https://github.com/jjo/prometheus/pull/1/commits/629009a8c165dc8d48d2a40a1561a65171ea0462
## PR Analysis

- 🎯 **Main theme:** This PR introduces new features and improvements to the Prometheus project. It includes updates to Go and Node.js versions, lint fixes, unit tests for native histograms, and changes to the promql and tsdb packages.
- 📝 **PR summary:** The PR includes a variety of changes. It updates the Go version to 1.21 and the Node.js version to the latest LTS release. It also fixes a linting issue related to the deprecation of `rand.Read`. The PR introduces unit tests for native histograms and makes changes to the promql and tsdb packages. Additionally, it includes a variety of other minor fixes and improvements.
- 📌 **Type of PR:** Enhancement
- 🧪 **Relevant tests added:** Yes
- 🔒 **Security concerns:** No security concerns found
## PR Feedback

- 💡 **General suggestions:** The PR includes a variety of changes, which makes it difficult to review and understand the impact of the changes. It would be better to split the changes into multiple PRs, each focusing on a specific change. For example, updating the Go version could be one PR, adding unit tests for native histograms could be another PR, and so on. This would make the review process easier and more efficient.


- **<details><summary> 🤖 Code feedback:**</summary>


  - **relevant file:** ``promql/engine.go``
   **suggestion:** It seems like the `SAMPLE_RATIO` case in the `aggregation` function is quite complex and could potentially be simplified or broken down into smaller functions for better readability and maintainability. [medium]
   **relevant line:** ['+	if op == parser.SAMPLE_RATIO {'](https://github.com/jjo/prometheus/pull/1/files#diff-ad2761bb162a2f12326b18f2714ef54f4cd476cfe3388f77b2b20c5ff82e4faaR2546)


  - **relevant file:** ``promql/parser/parse_test.go``
   **suggestion:** The `TestParseHistogramSeries` and `TestHistogramTestExpression` test functions are quite long and could potentially be broken down into smaller test functions. This would make the tests easier to read and maintain. [medium]
   **relevant line:** ['+func TestParseHistogramSeries(t *testing.T) {'](https://github.com/jjo/prometheus/pull/1/files#diff-e22dce06c487c41c75ba9e0069b1f3da8667e5314b56132cf7b93ca91b5829dcR3711)


  - **relevant file:** ``tsdb/db.go``
   **suggestion:** The `ForceHeadMMap` function is only intended for use in tests and benchmarks. It would be better to move this function to a `_test.go` file to clearly indicate its intended use and to prevent it from being part of the package's public API. [important]
   **relevant line:** ['+func (db *DB) ForceHeadMMap() {'](https://github.com/jjo/prometheus/pull/1/files#diff-495908409affb7da01cb15f2a89b405f105378984919d6149b5cdaaf5e275c7eR1801)

</details>


### How to use
> Tag me in a comment '@CodiumAI-Agent' and add one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
> **/improve [--extended]**: Suggest improvements to the code in the PR. Extended mode employs several calls, and provides a more thorough feedback. 
> **/ask \<QUESTION\>**: Pose a question about the PR.
> **/update_changelog**: Update the changelog based on the PR's contents.

>To edit any configuration parameter from **configuration.toml**, add --config_path=new_value
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 
>To list the possible configuration parameters, use the **/config** command.
